README.md

# vecfx: extra batteries for a vec of floats

[![Build Status](https://travis-ci.org/ybyygu/vecfx.svg?branch=master)](https://travis-ci.org/ybyygu/vecfx)
[![GPL3 licensed](https://img.shields.io/badge/license-GPL3-blue.svg)](./LICENSE)

Rust's ecosystem system is not perfect for scientific computation yet. To reduce
the growing pains, a safe choice is sticking to the standard types for public
interfacing. This crate provides the extra powers of Vec/Array/slice type for
simple computation, enabled by Rust's trait system.



CHANGELOG.md

# <span class="timestamp-wrapper"><span class="timestamp">[2020-02-18 Tue] </span></span> v0.1.0

-   fix doc typo


# <span class="timestamp-wrapper"><span class="timestamp">[2019-12-23 Mon] </span></span> v0.0.8

-   re-exports nalgebra package


# <span class="timestamp-wrapper"><span class="timestamp">[2019-12-10 Tue] </span></span> v0.0.7

-   expose Matrix3f type


# <span class="timestamp-wrapper"><span class="timestamp">[2019-12-06 Fri] </span></span> v0.0.6

-   add mean, var, std\_dev methods adopted from libtest.


# <span class="timestamp-wrapper"><span class="timestamp">[2019-05-10 Fri] </span></span> v0.0.5

-   add methods for vector projection/rejection


# <span class="timestamp-wrapper"><span class="timestamp">[2019-04-29 Mon] </span></span> v0.0.4

-   update nalgebra to v0.18


# <span class="timestamp-wrapper"><span class="timestamp">[2019-02-19 Tue] </span></span> v0.0.3

-   new methods for float array: \`imin\`, \`imax\`


# <span class="timestamp-wrapper"><span class="timestamp">[2019-01-09 Wed] </span></span> v0.0.2

-   new method \`vecdist\_squared\`



src/lib.rs
//! Backend for vector operations

// [[file:../vecfx.note::*imports][imports:1]]
use std::f64;
// imports:1 ends here

// [[file:../vecfx.note::*mods][mods:1]]
#[cfg(feature = "nalgebra")]
mod projection;
#[cfg(feature = "nalgebra")]
#[cfg(feature = "adhoc")]
mod rotation;

mod array;
mod iterator;
mod ord;
mod stats;
mod vector;
// mods:1 ends here

// [[file:../vecfx.note::*pub][pub:1]]
pub use crate::stats::*;

#[cfg(feature = "nalgebra")]
pub use projection::*;

pub use crate::vector::*;

#[cfg(feature = "nalgebra")]
pub use nalgebra;

pub use crate::iterator::FloatIterExt;

#[cfg(feature = "nalgebra")]
pub use crate::iterator::na::*;

#[cfg(feature = "nalgebra")]
// for using assert_relative_eq macro for Vector or Matrix
pub use approx;

pub use crate::ord::*;

pub use array::*;
// pub:1 ends here


src/ord.rs
// [[file:../vecfx.note::*impl][impl:1]]
use std::borrow::Borrow;

pub use ordered_float::{Float, OrderedFloat};

// Not very useful
trait IntoOrderedExt {
    fn into_ordered(self) -> Vec<OrderedFloat<f64>>;
}

impl<F, T> IntoOrderedExt for T
where
    T: Iterator<Item = F>,
    F: std::borrow::Borrow<f64>,
{
    /// Convert an iterator over floats into a vec of ordered floats
    fn into_ordered(self) -> Vec<OrderedFloat<f64>> {
        self.map(|x| OrderedFloat(*x.borrow())).collect()
    }
}

/// Provides method for sorting a vec of floats
pub trait SortByExt {
    fn sort_by_float(&mut self);
}

impl<F: Float> SortByExt for [F] {
    /// Sort by float numbers
    fn sort_by_float(&mut self) {
        self.sort_by_key(|x| OrderedFloat(*x));
    }
}
// impl:1 ends here

// [[file:../vecfx.note::*test][test:1]]
#[test]
fn test_float_ord() {
    let mut values = vec![1.0, -1.0, 2.0];
    let mut values_ordered = values.iter().into_ordered();
    values_ordered.sort();

    assert_eq!(values_ordered[0], -1.0);
    assert_eq!(values_ordered[2], 2.0);

    values.sort_by_float();
    assert_eq!(values[0], values_ordered[0].into());
    assert_eq!(values[1], values_ordered[1].into());
}
// test:1 ends here


src/array.rs
// [[file:../vecfx.note::*imports][imports:1]]
use crate::Float;
// imports:1 ends here

// [[file:../vecfx.note::*core][core:1]]
#[inline]
fn array_sub_generic<F, const N: usize>(x: [F; N], y: [F; N]) -> [F; N]
where
    F: Float + std::fmt::Debug,
{
    use std::convert::TryInto;
    let d: Vec<_> = x.iter().zip(y.iter()).map(|(a, b)| *a - *b).collect();
    d.try_into().unwrap()
}

#[inline]
fn array_add_generic<F, const N: usize>(x: [F; N], y: [F; N]) -> [F; N]
where
    F: Float + std::fmt::Debug,
{
    use std::convert::TryInto;
    let d: Vec<_> = x.iter().zip(y.iter()).map(|(a, b)| *a + *b).collect();
    d.try_into().unwrap()
}

/// Simple math for array
pub trait ArrayMathExt<F: Float> {
    fn array_add(self, other: Self) -> Self;
    fn array_sub(self, other: Self) -> Self;
    fn array_scale(self, value: F) -> Self;
}

impl<F, const N: usize> ArrayMathExt<F> for [F; N]
where
    F: Float + std::fmt::Debug,
{
    /// Adds two arrays. Returns x + y.
    fn array_add(self, other: Self) -> Self {
        array_add_generic(self, other)
    }

    /// Subtracts two arrays. Returns x - y.
    fn array_sub(self, other: Self) -> Self {
        array_sub_generic(self, other)
    }

    /// Scale one array with a value.
    fn array_scale(self, value: F) -> Self {
        self.map(|x| x * value)
    }
}
// core:1 ends here

// [[file:../vecfx.note::*test][test:1]]
#[test]
fn test_array_sub_add() {
    use crate::approx::*;

    let p1 = [1.0, 2.0, 3.0];
    let p2 = [1.1, 2.2, 3.3];
    let p3 = p2.array_sub(p1);
    let p4 = p2.array_add(p1);
    assert_relative_eq!(p3[0], 0.1, epsilon = 1e-4);
    assert_relative_eq!(p3[1], 0.2, epsilon = 1e-4);
    assert_relative_eq!(p3[2], 0.3, epsilon = 1e-4);
    assert_relative_eq!(p4[0], 2.1, epsilon = 1e-4);
    assert_relative_eq!(p4[1], 4.2, epsilon = 1e-4);
    assert_relative_eq!(p4[2], 6.3, epsilon = 1e-4);
    let p5 = p1.array_scale(2.0);
    assert_eq!(p5[0], 2.0);
    assert_eq!(p5[1], 4.0);
    assert_eq!(p5[2], 6.0);
}
// test:1 ends here


src/stats.rs
// stats
// abstracted from: https://github.com/rust-lang/libtest/blob/master/libtest/stats.rs

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*stats][stats:1]]
/// Trait that provides simple descriptive statistics on a univariate set of numeric samples.
pub trait StatsExt {
    /// Sum of the samples.
    ///
    /// Note: this method sacrifices performance at the altar of accuracy
    /// Depends on IEEE-754 arithmetic guarantees. See proof of correctness at:
    /// ["Adaptive Precision Floating-Point Arithmetic and Fast Robust Geometric
    /// Predicates"][paper]
    ///
    /// [paper]: http://www.cs.cmu.edu/~quake-papers/robust-arithmetic.ps
    fn sum(&self) -> f64;

    /// Minimum value of the samples.
    fn min(&self) -> f64;

    /// Maximum value of the samples.
    fn max(&self) -> f64;

    /// Arithmetic mean (average) of the samples: sum divided by sample-count.
    ///
    /// See: <https://en.wikipedia.org/wiki/Arithmetic_mean>
    fn mean(&self) -> f64;

    /// Variance of the samples: bias-corrected mean of the squares of the differences of each
    /// sample from the sample mean. Note that this calculates the _sample variance_ rather than the
    /// population variance, which is assumed to be unknown. It therefore corrects the `(n-1)/n`
    /// bias that would appear if we calculated a population variance, by dividing by `(n-1)` rather
    /// than `n`.
    ///
    /// See: <https://en.wikipedia.org/wiki/Variance>
    fn var(&self) -> f64;

    /// Standard deviation: the square root of the sample variance.
    ///
    /// Note: this is not a robust statistic for non-normal distributions. Prefer the
    /// `median_abs_dev` for unknown distributions.
    ///
    /// See: <https://en.wikipedia.org/wiki/Standard_deviation>
    fn std_dev(&self) -> f64;

    /// Index to the minimum value of the samples.
    fn imin(&self) -> usize;

    /// Index to the maximum value of the samples.
    fn imax(&self) -> usize;
}

impl StatsExt for [f64] {
    // FIXME #11059 handle NaN, inf and overflow
    fn sum(&self) -> f64 {
        let mut partials = vec![];

        for &x in self {
            let mut x = x;
            let mut j = 0;
            // This inner loop applies `hi`/`lo` summation to each
            // partial so that the list of partial sums remains exact.
            for i in 0..partials.len() {
                let mut y: f64 = partials[i];
                if x.abs() < y.abs() {
                    std::mem::swap(&mut x, &mut y);
                }
                // Rounded `x+y` is stored in `hi` with round-off stored in
                // `lo`. Together `hi+lo` are exactly equal to `x+y`.
                let hi = x + y;
                let lo = y - (hi - x);
                if lo != 0.0 {
                    partials[j] = lo;
                    j += 1;
                }
                x = hi;
            }
            if j >= partials.len() {
                partials.push(x);
            } else {
                partials[j] = x;
                partials.truncate(j + 1);
            }
        }
        let zero: f64 = 0.0;
        partials.iter().fold(zero, |p, q| p + *q)
    }

    fn min(&self) -> f64 {
        assert!(!self.is_empty());
        self.iter().fold(self[0], |p, q| p.min(*q))
    }

    fn max(&self) -> f64 {
        assert!(!self.is_empty());
        self.iter().fold(self[0], |p, q| p.max(*q))
    }

    fn mean(&self) -> f64 {
        assert!(!self.is_empty());
        self.sum() / (self.len() as f64)
    }

    fn var(&self) -> f64 {
        if self.len() < 2 {
            0.0
        } else {
            let mean = self.mean();
            let mut v: f64 = 0.0;
            for s in self {
                let x = *s - mean;
                v += x * x;
            }
            // N.B., this is _supposed to be_ len-1, not len. If you
            // change it back to len, you will be calculating a
            // population variance, not a sample variance.
            let denom = (self.len() - 1) as f64;
            v / denom
        }
    }

    fn std_dev(&self) -> f64 {
        self.var().sqrt()
    }

    fn imin(&self) -> usize {
        assert!(!self.is_empty());
        self.iter()
            .enumerate()
            .fold(0, |i, (j, q)| if self[i] > *q { j } else { i })
    }

    fn imax(&self) -> usize {
        assert!(!self.is_empty());
        self.iter()
            .enumerate()
            .fold(0, |i, (j, q)| if self[i] > *q { i } else { j })
    }
}
// stats:1 ends here

// test

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*test][test:1]]
#[cfg(test)]
mod tests {
    use super::*;
    use approx::*;

    #[test]
    fn test_stats_nan() {
        let xs = &[1.0, 2.0, std::f64::NAN, 3.0, 4.0];
        assert_eq!(xs.min(), 1.0);
        assert_eq!(xs.max(), 4.0);
        assert_eq!(xs.imin(), 0);
        assert_eq!(xs.imax(), 4);
        assert!(xs.sum().is_nan());
        assert!(xs.mean().is_nan());
    }

    #[test]
    fn test_stats() {
        let val = &[958.0000000000, 924.0000000000];
        assert_relative_eq!(val.sum(), 1882.0);
        assert_relative_eq!(val.min(), 924.0);
        assert_relative_eq!(val.max(), 958.0);
        assert_relative_eq!(val.mean(), 941.0);
        assert_relative_eq!(val.var(), 578.0);
        assert_relative_eq!(val.std_dev(), 24.0416305603, epsilon = 1e-6);
        assert_eq!(val.imin(), 1);
        assert_eq!(val.imax(), 0);
    }
}
// test:1 ends here


src/iterator.rs
// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*na matrix][na matrix:1]]
#[cfg(feature = "nalgebra")]
pub(crate) mod na {
    use crate::nalgebra as na;
    use crate::vector::*;
    use crate::Matrix3f;

    pub trait FloatIterToVector {
        fn to_vector(self) -> na::DVector<f64>;
    }

    pub trait FloatIterToMatrix {
        fn to_matrix(self) -> Vector3fVec;
    }

    impl<T, F> FloatIterToVector for T
    where
        T: Iterator<Item = F>,
        F: std::borrow::Borrow<f64>,
    {
        /// Convert an iterator over floats as Nalgebra column vector
        fn to_vector(self) -> na::DVector<f64> {
            let d: Vec<f64> = self.map(|x| *x.borrow()).collect();
            d.to_column_vector()
        }
    }

    impl<F, T> FloatIterToMatrix for T
    where
        T: Iterator<Item = F>,
        F: std::borrow::Borrow<[f64; 3]>,
    {
        /// Convert an iterator over float array as Nalgebra 3xN matrix
        fn to_matrix(self) -> Vector3fVec {
            let d: Vec<[f64; 3]> = self.map(|x| *x.borrow()).collect();
            d.to_matrix()
        }
    }

    #[test]
    fn test_float_iter_trait() {
        let d = vec![1f64, 2.0, 3.0];
        let _ = d.iter().to_vector();
        let _ = d.into_iter().to_vector();

        let d = vec![[1f64; 3], [2.0; 3], [3.0; 3]];
        let _ = d.iter().to_matrix();
        let _ = d.into_iter().to_matrix();
    }
}
// na matrix:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*impl][impl:1]]
use std::borrow::Borrow;
use std::f64;

// https://stackoverflow.com/questions/43921436/extend-iterator-with-a-mean-method
/// max/min for iterator over floats.
///
/// # Example
/// ```
/// use vecfx::*;
/// 
/// let values = vec![1.0, 2.1, -1.1];
/// assert_eq!(values.iter().float_max(), 2.1);
/// assert_eq!(values.iter().float_min(), -1.1);
/// 
/// let values = vec![1.2, 0.0/0.0, 2.2];
/// assert_eq!(values.iter().float_max(), 2.2);
/// assert_eq!(values.iter().float_min(), 1.2);
/// ```
///
pub trait FloatIterExt<F>: Iterator<Item = F>
where
    F: std::borrow::Borrow<f64>,
{
    /// Returns the maximum element of an iterator.
    ///
    /// # Panic
    ///
    /// * panics if iterator is empty.
    fn float_max(&mut self) -> f64 {
        let value = *self.next().expect("float max: empty iterator").borrow();
        self.fold(value, |a, b| a.max(*b.borrow()))
    }

    /// Returns the minimum element of an iterator.
    ///
    /// # Panic
    ///
    /// * panics if iterator is empty.
    fn float_min(&mut self) -> f64 {
        let value = *self.next().expect("float min: empty iterator").borrow();
        self.fold(value, |a, b| a.min(*b.borrow()))
    }
}

impl<F, T> FloatIterExt<F> for T
where
    T: Iterator<Item = F>,
    F: std::borrow::Borrow<f64>,
{
}
// impl:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*test][test:1]]
#[cfg(test)]
mod tests {
    use crate::*;

    #[test]
    fn test_float_ordering() {
        let values = vec![1.0, -1.0, 0.0 / 0.0, 0.5, 2.0];
        assert_eq!(values.iter().float_max(), values.max());
        assert_eq!(values.iter().float_min(), values.min());

        let values = vec![1.0, -1.0, 0.0 / 0.0, 0.5, 2.0];
        assert_eq!(values.iter().float_max(), values.max());
        assert_eq!(values.iter().float_min(), values.min());

        // test for compatibility
        let values = vec![1.0f64, 2.0, 0.0, -9.0, 0.0 / 0.0];
        assert_eq!(values.iter().cloned().float_max(), values.max());
        assert_eq!(values.iter().cloned().float_min(), values.min());
    }

    #[test]
    #[should_panic]
    fn test_float_min_max_empty() {
        let values: Vec<f64> = Vec::new();
        let _ = values.iter().float_max();
    }
}
// test:1 ends here


src/rotation.rs
// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*imports][imports:1]]
use crate::{VecFloatExt, Vector3fVec};

use nalgebra as na;
// imports:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*tests][tests:1]]
use crate::*;

use na::geometry::Rotation3;
use na::Unit;
use na::Vector3;

#[test]
fn test_rotation() {
    use approx::*;

    // C2H6
    let coords = vec![
        [-2.1524, 0.6849, 0.0000],
        [-1.7958, -0.3239, 0.0000],
        [-1.7957, 1.1893, -0.8737],
        [-3.2224, 0.6849, 0.0000],
        [-1.6391, 1.4108, 1.2574],
        [-1.9941, 2.4202, 1.2564],
        [-1.9973, 0.9075, 2.1311],
        [-0.5691, 1.4091, 1.2584],
    ];

    // 沿2-1-5角, 转动原子2, PI/4, 原子的新坐标为:
    // -1.5435571	0.0976385	0.6552523
    let coords_a2 = [-1.5435571, 0.0976385, 0.6552523];

    let p1 = Vector3::from(coords[0]);
    let p2 = Vector3::from(coords[1]);
    let p5 = Vector3::from(coords[4]);
    let p6 = Vector3::from(coords[5]);

    let v12 = p2 - &p1;
    let v15 = p5 - &p1;
    let normal_215 = v12.cross(&v15).normalize();

    // 求1-2-5角度
    let angle_215 = v12.angle(&v15);
    dbg!(angle_215/3.1415 * 180.0);

    // let axis = Vector3::y_axis();
    let angle = std::f64::consts::FRAC_PI_4;
    let rot = Rotation3::from_axis_angle(&Unit::new_normalize(normal_215), angle);
    let p1_prime = rot * v12 + p1;
    assert_relative_eq!(p1_prime, Vector3::from(coords_a2), epsilon = 1e-4);

    // 旋转5,6,7,8基团 PI/4
    let normal_215 = v15.cross(&v12).normalize();
    let rot = Rotation3::from_axis_angle(&Unit::new_normalize(normal_215), angle);
    let p6_prime = rot * (p6 - &p1) + p1;
    dbg!(p6_prime);



    // Point and vector being transformed in the tests.
    let n = Vector3::new(1.0, 0.1, 0.0);
    let t = Vector3::new(0.1, 1.0, 0.0);
    let normal = n.cross(&t).normalize();
    dbg!(normal);

    let rot = Rotation3::from_axis_angle(&Unit::new_normalize(normal), angle);

    // let rot = Rotation3::from_axis_angle(&vec.normalize(), angle);

    dbg!(rot * n);
}
// tests:1 ends here


src/projection.rs
// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*imports][imports:1]]
use crate::{VecFloatExt, Vector3fVec};

use nalgebra as na;
// imports:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*base][base:1]]
pub trait VectorProject {
    fn cosine_similarity(&self, vb: &Self) -> f64;

    fn scalar_projection(&self, vb: &Self) -> f64;

    fn vector_projection(&self, vb: &Self) -> Self;

    fn vector_rejection(&self, vb: &Self) -> Self;
}
// base:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*core][core:1]]
macro_rules! impl_vector_project {
    ($type_name:ty) => {
        impl VectorProject for $type_name {
            /// The cosine of two non-zero vectors `va` and `vb`
            fn cosine_similarity(&self, vb: &Self) -> f64 {
                self.normalize().dot(&vb.normalize())
            }

            /// the scalar projection of a vector a onto a vector b
            fn scalar_projection(&self, vb: &Self) -> f64 {
                self.cosine_similarity(vb) * &self.norm()
            }

            /// The vector component of `va` parallel to `vb`
            fn vector_projection(&self, vb: &Self) -> Self {
                self.scalar_projection(vb) * vb.normalize()
            }

            /// The vector component of `va` perpendicular to `vb`
            fn vector_rejection(&self, vb: &Self) -> Self {
                self - &self.vector_projection(vb)
            }
        }
    };
}

impl_vector_project!(na::DVector<f64>);
impl_vector_project!(Vector3fVec);
// core:1 ends here

// [[file:~/Workspace/Programming/gchemol-rs/vecfx/vecfx.note::*test][test:1]]
#[test]
fn test_vector_projection() {
    use approx::*;

    let va = [1.0; 3].to_vector();
    let vb = [1.1, 1.2, 1.3].to_vector();

    let x = va.cosine_similarity(&vb);
    assert_relative_eq!(x, 0.9976931918526477, epsilon=1e-4);

    let x = va.scalar_projection(&vb);
    assert_relative_eq!(x, 1.728055, epsilon=1e-4);

    let x = va.vector_projection(&vb);
    let e = [0.9124424 , 0.99539171, 1.07834101];
    for i in 0..3 {
        assert_relative_eq!(x[i], e[i], epsilon=1e-4);
    }

    let x = va.vector_rejection(&vb);
    let e = [ 0.0875576 ,  0.00460829, -0.07834101];
    for i in 0..3 {
        assert_relative_eq!(x[i], e[i], epsilon=1e-4);
    }
}
// test:1 ends here


src/vector.rs
// [[file:../vecfx.note::*imports][imports:1]]
#[cfg(feature = "nalgebra")]
use nalgebra as na;
// imports:1 ends here

// [[file:../vecfx.note::*types][types:1]]
#[cfg(feature = "nalgebra")]
/// 3xN matrix storing a list of 3D vectors
pub type Vector3fVec =
    na::Matrix<f64, na::U3, na::Dynamic, na::VecStorage<f64, na::U3, na::Dynamic>>;

#[cfg(feature = "nalgebra")]
/// A stack-allocated, 3-dimensional column vector.
pub type Vector3f = na::Vector3<f64>;

#[cfg(feature = "nalgebra")]
/// A stack-allocated, column-major, 3x3 square matrix
pub type Matrix3f = na::Matrix3<f64>;

// #[cfg(feature = "nalgebra")]
// /// A dynamically sized column vector
// pub type Vector = na::DVector<f64>;
// types:1 ends here

// [[file:../vecfx.note::*for Vec<f64>][for Vec<f64>:1]]
/// Abstracting simple vector based math operations
pub trait VecFloatExt {
    /// y += c*x
    fn vecadd(&mut self, x: &[f64], c: f64);

    /// vector dot product
    /// s = x.dot(y)
    fn vecdot(&self, other: &[f64]) -> f64;

    /// y = z
    fn veccpy(&mut self, x: &[f64]);

    /// y = -x
    fn vecncpy(&mut self, x: &[f64]);

    /// z = x - y
    fn vecdiff(&mut self, x: &[f64], y: &[f64]);

    /// y *= c
    fn vecscale(&mut self, c: f64);

    /// ||x||, L2 norm
    fn vec2norm(&self) -> f64;

    /// 1 / ||x||
    fn vec2norminv(&self) -> f64;

    /// d = ||a-b||
    fn vecdist(&self, other: &[f64]) -> f64 {
        self.vecdist_squared(other).sqrt()
    }

    /// d = ||a-b||^2
    fn vecdist_squared(&self, other: &[f64]) -> f64;

    #[cfg(feature = "nalgebra")]
    /// Create dynamically allocated column vector from self
    fn to_column_vector(&self) -> na::DVector<f64>;

    #[cfg(feature = "nalgebra")]
    /// Create dynamically allocated column vector from self
    fn to_vector(&self) -> na::DVector<f64> {
        self.to_column_vector()
    }
}

impl VecFloatExt for [f64] {
    /// y += c*x
    fn vecadd(&mut self, x: &[f64], c: f64) {
        for (y, x) in self.iter_mut().zip(x) {
            *y += c * x;
        }
    }

    /// s = y.dot(x)
    fn vecdot(&self, other: &[f64]) -> f64 {
        self.iter().zip(other).map(|(x, y)| x * y).sum()
    }

    /// y *= c
    fn vecscale(&mut self, c: f64) {
        for y in self.iter_mut() {
            *y *= c;
        }
    }

    /// y = x
    fn veccpy(&mut self, x: &[f64]) {
        for (v, x) in self.iter_mut().zip(x) {
            *v = *x;
        }
    }

    /// y = -x
    fn vecncpy(&mut self, x: &[f64]) {
        for (v, x) in self.iter_mut().zip(x) {
            *v = -x;
        }
    }

    /// z = x - y
    fn vecdiff(&mut self, x: &[f64], y: &[f64]) {
        for ((z, x), y) in self.iter_mut().zip(x).zip(y) {
            *z = x - y;
        }
    }

    /// ||x||, L2 norm
    fn vec2norm(&self) -> f64 {
        let n2 = self.vecdot(&self);
        n2.sqrt()
    }

    /// 1/||x||
    fn vec2norminv(&self) -> f64 {
        1.0 / self.vec2norm()
    }

    /// d = ||a-b||^2
    fn vecdist_squared(&self, other: &[f64]) -> f64 {
        self.iter().zip(other).map(|(a, b)| (a - b).powi(2)).sum::<f64>()
    }

    #[cfg(feature = "nalgebra")]
    fn to_column_vector(&self) -> na::DVector<f64> {
        na::DVector::from_column_slice(&self)
    }
}

#[test]
fn test_vec_math() {
    use approx::*;

    // vector scaled add
    let x = [1.0, 1.0, 1.0];
    let c = 2.;

    let mut y = [1.0, 2.0, 3.0];
    y.vecadd(&x, c);

    assert_eq!(3.0, y[0]);
    assert_eq!(4.0, y[1]);
    assert_eq!(5.0, y[2]);

    // vector dot
    let v = y.vecdot(&x);
    assert_eq!(12.0, v);

    // vector scale
    y.vecscale(2.0);
    assert_eq!(6.0, y[0]);
    assert_eq!(8.0, y[1]);
    assert_eq!(10.0, y[2]);

    // vector diff
    let mut z = y.clone();
    z.vecdiff(&x, &y);
    assert_eq!(-5.0, z[0]);
    assert_eq!(-7.0, z[1]);
    assert_eq!(-9.0, z[2]);

    // vector copy
    y.veccpy(&x);

    // y = -x
    y.vecncpy(&x);
    assert_eq!(-1.0, y[0]);
    assert_eq!(-1.0, y[1]);
    assert_eq!(-1.0, y[2]);

    // vector distance
    let x = [0.0, 0.0, 0.0];
    let y = [1.0, 1.0, 1.0];
    let r = x.vecdist(&y);
    assert_relative_eq!(r, y.vec2norm(), epsilon = 1e-4);
}

#[cfg(feature = "nalgebra")]
#[test]
fn test_vec_math_na() {
    // nalgebra
    let y = [-1.0; 3];
    let v = y.to_column_vector();
    assert_eq!(v.norm_squared(), 3.0);
}

/// View a flat slice as nested 3D array
///
/// # Panics
/// if the slice size is incorrect.
pub trait VecFloatAs3D {
    /// View `&[f64]` as `&[[f64; 3]]` without copying.
    fn as_3d(&self) -> &[[f64; 3]];

    /// View `&mut [f64]` as `&mut [[f64; 3]]` without copying.
    fn as_mut_3d(&mut self) -> &mut [[f64; 3]];
}

impl VecFloatAs3D for [f64] {
    fn as_3d(&self) -> &[[f64; 3]] {
        assert_eq!(0, self.len() % 3, "cannot view slice of length {} as &[[_; 3]]", self.len());
        unsafe { ::std::slice::from_raw_parts(self.as_ptr() as *const _, self.len() / 3) }
    }

    fn as_mut_3d(&mut self) -> &mut [[f64; 3]] {
        assert_eq!(0, self.len() % 3, "cannot view slice of length {} as &[[_; 3]]", self.len());
        unsafe { ::std::slice::from_raw_parts_mut(self.as_ptr() as *mut _, self.len() / 3) }
    }
}
// for Vec<f64>:1 ends here

// [[file:../vecfx.note::*for Vec<\[f64; 3\]>][for Vec<[f64; 3]>:1]]
pub trait VecFloat3Ext {
    /// Return a 1-D array, containing the elements of 3xN array
    fn ravel(&self) -> Vec<f64> {
        self.as_flat().to_vec()
    }

    /// View as a flat slice
    fn as_flat(&self) -> &[f64];

    /// View of mut flat slice
    fn as_mut_flat(&mut self) -> &mut [f64];

    #[cfg(feature = "nalgebra")]
    /// Create a 3xN matrix of nalgebra from self
    fn to_matrix(&self) -> Vector3fVec;

    #[cfg(feature = "nalgebra")]
    #[cfg(feature = "adhoc")]
    /// Return distance matrix
    fn distance_matrix(&self) -> na::DMatrix<f64>;
}

impl VecFloat3Ext for [[f64; 3]] {
    /// View as a flat slice
    fn as_flat(&self) -> &[f64] {
        unsafe { ::std::slice::from_raw_parts(self.as_ptr() as *const _, self.len() * 3) }
    }

    /// View of mut flat slice
    fn as_mut_flat(&mut self) -> &mut [f64] {
        unsafe { ::std::slice::from_raw_parts_mut(self.as_mut_ptr() as *mut _, self.len() * 3) }
    }

    #[cfg(feature = "nalgebra")]
    /// Create a 3xN matrix of nalgebra from self
    fn to_matrix(&self) -> Vector3fVec {
        let r = self.as_flat();
        Vector3fVec::from_column_slice(r)
    }

    /// Return distance matrix
    #[cfg(feature = "nalgebra")]
    #[cfg(feature = "adhoc")]
    fn distance_matrix(&self) -> na::DMatrix<f64> {
        let n = self.len();

        let mut distances = na::DMatrix::zeros(n, n);
        for i in 0..n {
            for j in 0..i {
                let d = self[i].vecdist(&self[j]);
                distances[(i, j)] = d;
                distances[(j, i)] = d;
            }
        }

        distances
    }
}

#[test]
fn test_vecf3() {
    use approx::*;

    #[cfg(feature = "nalgebra")]
    {
        let a = vec![1.0, 2.0, 3.0];
        let x = a.to_vector();
        assert_relative_eq!(x.norm(), a.vec2norm(), epsilon = 1e-3);
    }

    let positions = [
        [-0.131944, -0.282942, 0.315957],
        [0.40122, -1.210646, 0.315957],
        [-1.201944, -0.282942, 0.315957],
        [0.543331, 0.892036, 0.315957],
        [0.010167, 1.819741, 0.315957],
        [1.613331, 0.892036, 0.315957],
    ];

    #[cfg(feature = "nalgebra")]
    {
        let n = positions.as_flat().vec2norm();
        let m = positions.to_matrix();
        assert_relative_eq!(n, m.norm(), epsilon = 1e-4);
    }

    let x = positions.ravel();
    assert_eq!(positions.len() * 3, x.len());

    let flat = positions.as_flat();
    assert_eq!(18, flat.len());

    let mut positions = positions.clone();
    let mflat = positions.as_mut_flat();
    mflat[0] = 0.0;
    assert_eq!(0.0, positions[0][0]);
}
// for Vec<[f64; 3]>:1 ends here

// [[file:../vecfx.note::*for Vec<\[f64; 3\]>][for Vec<[f64; 3]>:2]]
#[cfg(feature = "nalgebra")]
impl VecFloatAs3D for Vector3fVec {
    fn as_3d(&self) -> &[[f64; 3]] {
        assert_eq!(
            0,
            self.len() % 3,
            "cannot view Matrix of length {} as &[[_; 3]]",
            self.len()
        );

        self.as_slice().as_3d()
    }

    fn as_mut_3d(&mut self) -> &mut [[f64; 3]] {
        assert_eq!(
            0,
            self.len() % 3,
            "cannot view Matrix of length {} as &[[_; 3]]",
            self.len()
        );

        self.as_mut_slice().as_mut_3d()
    }
}

#[test]
fn test_as_3d() {
    let v = [1., 2., 3.];
    let p = v.as_3d();
    assert_eq!(&[[1., 2., 3.]], p);

    let mut v = vec![1., 2., 3., 4., 5., 6.];
    let p = &mut v.as_mut_3d();
    assert_eq!(p, &mut [[1., 2., 3.], [4., 5., 6.],]);
}

#[test]
#[cfg(feature = "nalgebra")]
fn test_as_3d_na() {
    let p = [1., 2., 3.];
    let mut m = p.as_3d().to_matrix();
    let _ = m.norm();

    let mut v = vec![[1., 2., 3.], [4., 5., 6.]].to_matrix();
    let mp = v.as_mut_3d();
    assert_eq!(mp, &mut [[1., 2., 3.], [4., 5., 6.],]);

    mp[0][0] = 1.1;
    assert_eq!(1.1, v[(0, 0)]);
}
// for Vec<[f64; 3]>:2 ends here

// [[file:../vecfx.note::*test][test:1]]
#[cfg(feature = "nalgebra")]
#[test]
fn test_vector3f() {
    use approx::*;

    let a = Vector3f::from([1.0, 2.0, 3.0]);
    assert_eq!(a.x, 1.0);
    assert_eq!(a.y, 2.0);
    assert_eq!(a.z, 3.0);
    assert_eq!(a[0], 1.0);
    assert_eq!(a[1], 2.0);
    assert_eq!(a[2], 3.0);

    let b = Vector3f::from([1.0, 2.0, 3.0]);
    let c = a + b;
    assert_relative_eq!(c.sum(), 12.0);
}
// test:1 ends here
